<center> <h2> DS 3000 - Fall 2021</h2> </center>
<center> <h3> DS Report </h3> </center>


<center> <h3>Predicting MyAnimeList Users' Anime Ratings</h3> </center>
<center><h4>Caitlin Chao, Rachael Cheung, Michelle Lim</h4></center>


<hr style="height:2px; border:none; color:black; background-color:black;">

#### Executive Summary:

In our Introduction, we review our overall topic of Anime recommendations as well as why Anime recommendations is significant. We also go over various questions that will be tackled by the rest of the project. For our Method, we discuss the existing datasets we found through Kaggle and the feature variables and target variable we decided to work with. We also decided that we will consider k Nearest Neighbors, Decision Tree, and Naive Bayes as machine learning algorithms. In Results, we perform data cleaning and wrangling so that the data can be analyzed, as well as split our data into testing and training sets to do further analysis. We also provide visualizations of our data. For the Discussion portion of this project, we compare the results we found by using the three machine learning algorithms, as results after hyperparameter tuning and feature selection. We also consider any ethical implications and future work we are interested in with this project in general.


<hr style="height:2px; border:none; color:black; background-color:black;">

## Outline
1. <a href='#1'>INTRODUCTION</a>
2. <a href='#2'>METHOD</a>
3. <a href='#3'>RESULTS</a>
4. <a href='#4'>DISCUSSION</a>

<a id="1"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 1. INTRODUCTION


<h4><b>Problem Statement</b></h4>

Anime viewers need better recommendations now that streaming platforms have become increasingly popular due to COVID-19 and quarantining policies. This can be done if there were a way to predict how a user would feel about an anime based on their previous reviews on other animes. Our project is to determine the score a user would give an anime based on multiple factors based on the previous ratings they have given, or rather, how much a user would like a certain anime they have not yet seen. Our results may vary if the target user's available data is too widespread, or if the amount of data given to the system is too small. Through this project, we want to learn how we could predict a user's anime rating based on previous data about the animes they have watched.


<h4><b>Significance of the Problem</b></h4>

In times of COVID-19, the use of streaming services has become widespread as more and more people seek out entertainment from their homes. In particular, aspects of Japanese culture like anime and manga have gotten more popular thanks to social media apps like Tiktok making these topics skyrocket in popularity. With that in mind, finding the next series or movie to watch is an issue we want to tackle in our project. Insights from this project can help us predict how a user would feel about an anime based on multiple factors, which can be used in turn to support a recommendation system for users.

<h4><b>Questions</b></h4>

Given the aforementioned problem and its importance, we set out to tackle the following questions:

>**Can we predict a user’s ratings for an anime based on some characteristics of the user and the anime series?**

>**Can we predict a user’s ratings for an anime based on their birthday and the anime’s genre?**

These two questions take into account features such as the user's birthday and the anime in question's genre. The user's birthday corresponds to their age, which we can deduce from the year their birthdate is in. Using this data, we will determine if there is correlation between a user's age, and the anime's genre. We want to see if, in general, an older population is more drawn to a certain genre compared a younger population.

><b>Would k Nearest Neighbors, Decision Tree Classifier, or Gaussian Naive Bayes be more effective in predicting the score a user would rate an anime based on the user's age and gender and the anime series's popularity, number of episodes and rank?</b>

This question aims to compare the effectiveness and accuracies of various machine learning algorithms in predicting/classifying our target variable (overall score) from our feature variables. By using these algorithms on our data, we will be able to tell numerically which is more effective.

<a id="2"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 2. METHOD

### 2.1. Data Acquisition

Our data was obtained via [Kaggle](https://www.kaggle.com/marlesson/myanimelist-dataset-animes-profiles-reviews) (if URL doesn't show: https://www.kaggle.com/marlesson/myanimelist-dataset-animes-profiles-reviews). The original dataset had 3 different spreadsheets containing information on MyAnimeList listings - one for anime, one for user profiles, and one for user reviews. We ended up merging the 3 datasets together to create one continuous spreadsheet.

For the anime spreadsheet, the data had the following variables: anime ID, title, its synopsis, its genres, whether it had finished airing, # of episodes, # of users who had watched it, its popularity and ranking, overall score, an image URL for the anime, and a link to its listing on MyAnimeList. For the purpose of this project, we disregarded the last 2 columns potential feature variables. This spreadsheet essentially contained information about all the anime listed on MyAnimeList.

The user profile spreadsheet had 5 columns for user profiles: their username, gender, birthday, favorite anime, and a link to their profile. For the purpose of this project, we disregarded profile links. This spreadsheet helps us connect users tracking anime on MyAnimeList to the anime they enjoyed the most, providing the basis of what we eventually plan to use the data for - determining whether a given user would enjoy a given anime.

Finally, the anime reviews spreadsheet had 7 columns: review ID, reviewer, anime ID, the review itself, the overall score (out of 10, with 1 being the lowest and 10 being the highest), score breakdown (1-10, factoring in the overall rating, story, animation, sound, character, and user enjoyment), and a link to the review. For the purpose of this project, we disregarded the last column. The review spreadsheet provides a further breakdown in order to help us gain insight into users' tastes - using our anime spreadsheet, we can now see what parts of an anime a user likes as well as detect any trends within a user's ratings to discern if they have a specific preference for a certain anime genre.

Once we combined all 3, the final dataset that we used for our project ended up having 10 feature variables and 2 target variables. Our feature variables were user profile/birthday/gender/favorites and an anime's ID/title/genre/episode count/popularity/ranking. Our outcome variables were the categorical scores that a user gave to a given anime, and the overall score they subsequently gave. This single CSV file had over 75000 rows.

--- **DATA STORED AT** https://raw.githubusercontent.com/caitchao/ds3000-project-data-anime/main/anime_data.csv ---

Data Wrangling Notebook: https://github.com/caitchao/ds3000-project-data-anime/blob/main/fp2_datawrangling.ipynb 

### 2.2. Data Analysis

Our 10 feature variables are the following:
|Features|
----------
|user profile|
|user birthday|
|user gender|
|user's favorite anime|
|anime's ID on MyAnimeList|
|anime title|
|anime genre|
|anime episode count|
|anime popularity on MyAnimeList|
|anime ranking on MyAnimeList|

Our 2 outcome variables were the categorical scores that a user gave to a given anime, and the overall score they subsequently gave it.

These feature variables are important predictors because they help us establish a connection between user demographics and the subsequent ratings that they give an anime. From user data, we can figure out if there is a certain genre that a user enjoys, or if there is a certain age or gender group among MyAnimeList users that will tend to enjoy an anime more than other users who do not fall into this group. Similarly, with the feature variables that have to do with an anime also serve as important predictors. We can determine if shorter/longer anime have higher or lower reviews, and the popularity/ranking on MyAnimeList can reveal trends in how a user decides what anime to watch next. Similarly, one genre may have higher rankings overall because it is less niche and caters to a broader demographic. The title of an anime can also provide a connection to increased viewership and rating if the anime is part of a series that spans multiple seasons.

This is a supervised machine learning problem because we have a set of feature and outcome variables in our training set that we can use to predict a user's rating of an anime based on the algorithm's classification of the variables and associated rating score. There are enough feature variables that we can use to ensure that our data classification is not too broad, but we would not be overfitting data points either. Based on all the feature variables, we will be able to predict the rating a user would give a certain anime thanks to the large amount of data we will be training and an algorithm to classify all of it. 

We plan on considering using the following 3 algorithms for our project:


1.   **k Nearest Neighbors**
    <br>k Nearest Neighbors operates under the assumption that similar points of
    data exist in close proximity, essentially, birds of a feather flock together. kNN looks at all the specific points closest to the supplied point to predict, and counts the "votes" for the classification that appears the most. Since our data is not sparse, it is reasonable to assume that it would work well with our dataset since it would help us classify a supplied point and there is a lot of data to work with, hence ensuring an accurate classification with the calculated predictors and patterns within the data. </br>

2.   **Decision Tree**
    <br>Decision trees assume that the entire training set is the "root" at the start of the tree/algorithm, and proceeds to categorize data in the set based on "tests" of whether data has a specific attribute or not. With 10 feature variables, we would be able to classify each data point easily because of the in-depth features we are using for the training set. Any nodes with similar attributes would be placed together. The algorithm would learn based on the decisions/categories the algorithm has made and subsequently be able to predict how well a new anime show would fare in terms of user review.</br>

3.   **Naive Bayes**
    <br>Naive Bayes operates under the assumption that all features are equally prioritized in the classification system and all features are statistically independent. Subsequently, the algorithm uses probabilistic classifiers to partition the data; thus, similar points end up in the same classifier. Based on the probability that a supplied data point to predict falls into a certain category, we can predict which category that a new data point would fall into. Since we have many data points that share similar values in terms of one (or more) of our feature variables, using Naive Bayes would help us split the data up into groups of data points that are similar to one another.</br>


<a id="3"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 3. RESULTS

### 3.1. Data Wrangling


In [ ]:
#read data into df
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/caitchao/ds3000-project-data-anime/main/anime_data.csv', index_col=0)
df

profile  ... overall_score
0        DesolatePsyche  ...             8
1             kokuborou  ...            10
2            TheGostoso  ...             8
3       PeterFromRussia  ...             8
4             agendator  ...             4
...                 ...  ...           ...
74550             Aljas  ...             8
74551  BlackSwordsman09  ...            10
74552      souper_toast  ...             9
74553        Funnysmily  ...             8
74554         Rikudo134  ...             9

[74555 rows x 12 columns]



> These functions will be used to map onto certain columns of the dataframe for easier manipulation of the features.



In [ ]:
from datetime import datetime, date
#given a date of birth in the MM/DD/YYYY format the function calculates
#the age based on the date
def date_to_age(born):
    try:
        born = datetime.strptime(born, "%m/%d/%Y").date()
    except:
        return float("NaN")
    today = date.today()
    return today.year - born.year - ((today.month, today.day) < (born.month, born.day))

import string
#given a string representation of a list of genres, the function returns
#the first genre (main genre) of the list as a string
def get_dom_genre(s):
    for char in string.punctuation:
        if char != "-":
            s = s.replace(char, ' ')
    l = s.split()
    if len(l) > 0:
        return l[0]
    else:
        return float("NaN")

Functions get mapped onto columns, and column names are changed appropriately to prepare for feature preprocessing. 

In [ ]:
#map column for birthday to age
df["birthday"]=df["birthday"].map(date_to_age)
df = df.rename(columns={"birthday": "age"})

#map column for genre list to main genre
df["genre"]=df["genre"].map(get_dom_genre)

df = df.dropna()
df

profile  ...  overall_score
0        DesolatePsyche  ...              8
1             kokuborou  ...             10
2            TheGostoso  ...              8
3       PeterFromRussia  ...              8
4             agendator  ...              4
...                 ...  ...            ...
74550             Aljas  ...              8
74551  BlackSwordsman09  ...             10
74552      souper_toast  ...              9
74553        Funnysmily  ...              8
74554         Rikudo134  ...              9

[74089 rows x 12 columns]

Now we one-hot encode the gender and genre, both which are both categorical variables, to represent them in terms of 0s and 1s. We do this to make the cateogircal variables easier to work with, since numerical values are easier to work with. 

In [ ]:
#categorical columns to one-hot encode
features_a = df.loc[:, ['gender', 'genre']]

from sklearn.preprocessing import OneHotEncoder

#create the encoder
encoder = OneHotEncoder(sparse = False)

#encode the given columns
encoded_df = encoder.fit_transform(features_a)

#create dataframe with encoded columns
features_a = pd.DataFrame(encoded_df, columns = encoder.get_feature_names_out(["gender", "genre"]))

features_a

gender_Female  gender_Male  ...  genre_Thriller  genre_Vampire
0                0.0          1.0  ...             0.0            0.0
1                0.0          1.0  ...             0.0            0.0
2                0.0          1.0  ...             0.0            0.0
3                0.0          1.0  ...             0.0            0.0
4                0.0          1.0  ...             0.0            0.0
...              ...          ...  ...             ...            ...
74084            0.0          1.0  ...             0.0            0.0
74085            0.0          1.0  ...             0.0            0.0
74086            0.0          1.0  ...             0.0            0.0
74087            0.0          0.0  ...             0.0            0.0
74088            0.0          1.0  ...             0.0            0.0

[74089 rows x 41 columns]

Next, we use the standard scaler on three of the variables-- episodes, popularity and rank. This is to standardize the values in these columns so that the mean is 0 and standard deviation is 1. We do this so that the machine learning model does not prioritize one variable over another. 

The profile, favorite_anime, uid, and title feature variables are not going to be used for training because they are arbitrary. The scores target variable column is also not going to be used because it provides multiple categories for target variables, which all can be gauged through the overall_score column.

In [ ]:
# columns that we want to apply standard scaler on
features_b = df.loc[:, ['episodes', 'popularity', 'rank']]

from sklearn.preprocessing import StandardScaler

#create the scaler
scaler = StandardScaler()

#fit the scaler to the training data(certain features only)
scaler.fit(features_b) 

#transform features based on the (same) scaler
features_b = scaler.transform(features_b) 
features_b = pd.DataFrame(features_b, columns = scaler.get_feature_names_out(['episodes', 'popularity', 'rank']))

features_b

episodes  popularity      rank
0     -0.220944   -0.215518 -0.873109
1     -0.220944   -0.215518 -0.873109
2     -0.220944   -0.215518 -0.873109
3     -0.220944   -0.215518 -0.873109
4     -0.220944   -0.215518 -0.873109
...         ...         ...       ...
74084 -0.486784    1.289013 -0.021733
74085 -0.269279    0.682904  0.103257
74086  4.153324    2.028557  0.202794
74087  0.431571    2.556075  0.274216
74088 -0.486784    1.116000  0.251422

[74089 rows x 3 columns]

We now put together the pre-processed variables and original variables together with the target feature. 

In [ ]:
#putting data samples together after preprocessing
age_df = df.loc[:, ['age']]
scores_df = df.loc[:, ['overall_score']]
df_final = pd.concat([age_df, features_a, features_b, scores_df], axis=1)

df_final

age  gender_Female  gender_Male  ...  popularity      rank  overall_score
0       27.0            0.0          1.0  ...   -0.215518 -0.873109            8.0
1       29.0            0.0          1.0  ...   -0.215518 -0.873109           10.0
2       21.0            0.0          1.0  ...   -0.215518 -0.873109            8.0
3        0.0            0.0          1.0  ...   -0.215518 -0.873109            8.0
4       26.0            0.0          1.0  ...   -0.215518 -0.873109            4.0
...      ...            ...          ...  ...         ...       ...            ...
74550   28.0            NaN          NaN  ...         NaN       NaN            8.0
74551  116.0            NaN          NaN  ...         NaN       NaN           10.0
74552   27.0            NaN          NaN  ...         NaN       NaN            9.0
74553   26.0            NaN          NaN  ...         NaN       NaN            8.0
74554   20.0            NaN          NaN  ...         NaN       NaN            9.0

[74550 rows x 46 columns]

We then perform data cleaning on this final dataframe by removing any null values and rows with irregular data (for example if age of a user is less than or equal to 0).

In [ ]:
#data cleaning
cleaned_df = df_final.dropna()
cleaned_df = cleaned_df.reset_index()
cleaned_df = cleaned_df.drop(columns=["index"])
cleaned_df = cleaned_df[cleaned_df["age"]>0]

cleaned_df

age  gender_Female  gender_Male  ...  popularity      rank  overall_score
0      27.0            0.0          1.0  ...   -0.215518 -0.873109            8.0
1      29.0            0.0          1.0  ...   -0.215518 -0.873109           10.0
2      21.0            0.0          1.0  ...   -0.215518 -0.873109            8.0
4      26.0            0.0          1.0  ...   -0.215518 -0.873109            4.0
5      24.0            0.0          1.0  ...   -0.215518 -0.873109           10.0
...     ...            ...          ...  ...         ...       ...            ...
73623  20.0            0.0          1.0  ...    1.289013 -0.021733            6.0
73624  31.0            0.0          1.0  ...    0.682904  0.103257            6.0
73625  29.0            0.0          1.0  ...    2.028557  0.202794            6.0
73626  36.0            0.0          0.0  ...    2.556075  0.274216            1.0
73627  26.0            0.0          1.0  ...    1.116000  0.251422            9.0

[62495 rows x 46 columns]

After splitting our dataframe into the feature variables and the target variable, we use train/test split to randomly split our dataframe into the training and testing sets. 

In [ ]:
#split columns into features and target
features = cleaned_df.drop("overall_score", axis=1)
target = cleaned_df["overall_score"]

In [ ]:
features.shape

(62495, 45)

In [ ]:
target.shape

(62495,)

In [ ]:
from sklearn.model_selection import train_test_split

#split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(features, target, random_state=3000)

### 3.2. Data Exploration


**Histogram of Overall Scores**

This figure shows that the most frequent user ratings of overall scores for a particular anime series were 8's, 9's, or 10's. This is important for us to know as we are analyzing the data because we will know what to expect as a user's general rating. It gives us an idea of what types of scores are more significant within the dataset as a whole.

In [ ]:
import plotly.express as px 
fig = px.histogram(df, x="overall_score", title = 'Histogram of Overall Scores', 
                   labels = {'overall_score':'Overall Score', 'count':'Number of Scores'}, 
                   color_discrete_sequence=['bisque'], width = 1000, template='plotly_white')

#fig.update_xaxes(range=[0, 11])
#fig.update_layout(bargap=0.1)
#fig.show()

<a href="https://ibb.co/M57HccD"><img src="https://i.ibb.co/rfHhxxG/newplot.png" alt="newplot" border="0" width="800"></a>

**Relationship between number of user age and anime series popularity**

This visualization is a scatter plot graph that is categorized by the overall score that the user rated an anime.  This shows the distribution and gives an idea of how the age of a user and the popularity of an anime series may impact the score the user gives to the series.

In [ ]:
import plotly.express as px

fig = px.scatter(df, x="age", y="popularity", color = "overall_score",
                    template="ggplot2",
                    labels = {"age": "User Age", "popularity": "Popularity", "overall_score": "Score"},
                    title='Relationship between User Age and Anime Series Popularity')

#fig.update_xaxes(range=[0, 95])
#fig.show()

<a href="https://ibb.co/920g8YH"><img src="https://i.ibb.co/WFX5xKv/newplot-5.png" alt="newplot-5" border="0" width="1000"></a>

**Relationship between number of episodes and anime series popularity**

From this graphic, we can conclude that animes with more episodes tend to have less popularity than other anime series, and less scores in general which could be due to the fact that viewers are more likely to watch an anime series all the way through if it is shorter.

In [ ]:
import plotly.express as px

fig = px.scatter(df, x="episodes", y="popularity", color = "overall_score",
                    template="seaborn",
                    labels = {"episodes": "Number of Episodes", "popularity": "Popularity",  "overall_score": "Score"},
                    title='Relationship between Number of Episodes and Anime Series Popularity')

#fig.update_xaxes(range=[0, 200])
#fig.show()

<a href="https://ibb.co/XxxC9w9"><img src="https://i.ibb.co/3mmCwJw/newplot-6.png" alt="newplot-6" border="0" width="1000"></a>

### 3.3. Model Training


In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB

from sklearn.metrics import accuracy_score, f1_score

# dictionary contains three classifiers
classifiers = {"k-Nearest Neighbor": KNeighborsClassifier(), 
              "Decision Tree": DecisionTreeClassifier(), 
              "Gaussian Naive Bayes": GaussianNB()}

We chose to use accuracy and f1 score metrics to evaluate our algorithm performance. Accuracy can be used to determine how well the algorithm classifies, and the f1 score can be used to gague the classifiers' exactness and completeness. With both of these metrics, we can get a sense of the model's effectiveness by evaluating the scores based on True/False Positives

In [ ]:
# iterate through dictionary of classifiers
for name, clf in classifiers.items():
    
    #create the model by fitting the training data
    clf.fit(X=X_train, y=y_train)
    
    # making predictions
    pred = clf.predict(X_test)
    
    #prediction accuracy
    accuracy = accuracy_score(y_test, pred)
    fone_score = f1_score(y_test, pred, average='weighted')

    print(f"{name}:")
    print("\tAccuracy score on the test data: ", accuracy)
    print("\tF1 score on the test data: ", fone_score,"\n")

k-Nearest Neighbor:
	Accuracy score on the test data:  0.18541986687147977
	F1 score on the test data:  0.18757798821714072 

Decision Tree:
	Accuracy score on the test data:  0.1906041986687148
	F1 score on the test data:  0.19449872022858408 

Gaussian Naive Bayes:
	Accuracy score on the test data:  0.033154121863799284
	F1 score on the test data:  0.021385372234501786 



Next, we will try to perform feature selection using Recursive Feature Selection to select a subset of our original features and train the models with the selected features, so we can see if we can achieve the same using a fewer number of features. However, since we used one-hot encoding for two of our categorical variables, we will exclude them from this feature selection test, and try to feature select three variables from the age, episodes, rank, and popularity variables. 

In [ ]:
X_train_toselect = X_train[["age", "episodes", "rank", "popularity"]]
X_test_toselect = X_test[["age", "episodes", "rank", "popularity"]]

In [ ]:
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeRegressor

def RFE_feature_selection():

    # Create a RFE object
    select = RFE(DecisionTreeRegressor(random_state=3000), n_features_to_select = 3)

    # Fit the data to the RFE
    select.fit(X_train_toselect, y_train)

    # transform the scaled data with the selected features
    X_train_selected = select.transform(X_train_toselect)
    X_test_selected = select.transform(X_test_toselect)

    column_names = X_train_toselect.columns.values

    # output the selected features where the support value is True 
    print("Selected features after RFE:")
    for x in range(len(select.support_)):
      if select.support_[x] == True:
        print(column_names[x])

    return X_train_selected, X_test_selected 

In [ ]:
X_train_selected, X_test_selected = RFE_feature_selection()

Selected features after RFE:
age
rank
popularity


In [ ]:
# iterate through dictionary of classifiers
for name, clf in classifiers.items():
    
    #create the model by fitting the training  data
    clf.fit(X=X_train_selected, y=y_train)
    
    # making predictions
    pred = clf.predict(X_test_selected)
    
    #prediction accuracy
    accuracy = accuracy_score(y_test, pred)
    fone_score = f1_score(y_test, pred, average='weighted')

    print(f"{name}:")
    print("\tAccuracy score on the test data: ", accuracy)
    print("\tF1 score on the test data: ", fone_score,"\n")

k-Nearest Neighbor:
	Accuracy score on the test data:  0.18593189964157705
	F1 score on the test data:  0.18843740203001227 

Decision Tree:
	Accuracy score on the test data:  0.19898873527905786
	F1 score on the test data:  0.20264444361044853 

Gaussian Naive Bayes:
	Accuracy score on the test data:  0.19316436251920122
	F1 score on the test data:  0.12768617785821113 



From these results, we can see that using three of the features selcted through Recursive Feature Selection analysis produces similar or even enhanced results to the original set of features used, with higher accuracy and f1 scores.

### 3.4. Model Optimization


We would now like to tune our models using hyperparameters that are appropriate for each model. From the above data, we decided to proceed with Decision Tree model and the Gaussian Naive Bayes model, which had higher accuracy rates than k-Nearest Neighbors. We use hyperparameter tuning in an attempt to minimize overfitting, or matching too closely to the training set. We try to use hyperparameter tuning by fitting to our original feature variables and also to our selected features, in an attempt to optimize performance. We encountered a UserWarning when using cross validation in our GridSearchCV, but this did not interfere with obtaining results.

In [ ]:
# import GridSearchCV and fit GridSearchCV
from sklearn.model_selection import GridSearchCV

# possible values for each hyperparameter
param_grid = {"max_depth":[10, 100, 1000, 10000, 10000],
             "max_leaf_nodes": [10, 100, 1000, 10000, 100000]}

# apply Grid Search with decision tree classifier and evaluate performance of parameters with 5-fold evaluation
grid_search = GridSearchCV(DecisionTreeClassifier(), param_grid, cv=5)

#fit the grid search object on the training data 
grid_search.fit(X=X_train, y=y_train)

report = ''
# result of grid search
report += "Decision Tree cross-validation\n"
report += f"Best Cross-validation score:  {grid_search.best_score_}\n" 
report += f"Best parameters:  {grid_search.best_params_}\n"

# the performance of the best found parameters on the training 
report += f"Training set score with best parameters:  {grid_search.score(X_train, y_train)}\n"
    
print (report)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:680: UserWarning:

The least populated class in y has only 1 members, which is less than n_splits=5.



Decision Tree
Best parameters:  {'max_depth': 100, 'max_leaf_nodes': 100}
Training set score with best parameters:  0.2517548164110004



In [ ]:
# possible values for each hyperparameter
param_grid = {"max_depth":[10, 100, 1000, 10000, 10000],
             "max_leaf_nodes": [10, 100, 1000, 10000, 100000]}

# apply Grid Search with decision tree classifier and evaluate performance of parameters with 5-fold evaluation
grid_search = GridSearchCV(DecisionTreeClassifier(), param_grid, cv=5)

#fit the grid search object on the training data after feature seletion
grid_search.fit(X=X_train_selected, y=y_train)

report = ''
# result of grid search
report += "Decision Tree Cross Validation with selected features\n"
report += f"Best Cross-validation score:  {grid_search.best_score_}\n" 
report += f"Best parameters:  {grid_search.best_params_}\n"

# the performance of the best found parameters on the training 
report += f"Training set score with best parameters:  {grid_search.score(X_train_selected, y_train)}\n"
    
print (report)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:680: UserWarning:

The least populated class in y has only 1 members, which is less than n_splits=5.



Decision Tree
Best parameters:  {'max_depth': 100, 'max_leaf_nodes': 100}
Training set score with best parameters:  0.249173262785091



In [ ]:
# possible values for each hyperparameter
param_grid = {"var_smoothing":[1e-09, 1e-08, 1e-07, 1e-06, 1e-05, 1e-04, 1e-03, 0.01, 0.1, 1, 10, 100]}

# apply Grid Search with Gaussian Naive Bayes model and evaluate performance of parameters with 5-fold evaluation
grid_search = GridSearchCV(GaussianNB(), param_grid, cv=5)

#fit the grid search object on the training data 
grid_search.fit(X=X_train, y=y_train)

report = ''
# result of grid search
report += "Gaussian Naive Bayes cross-validation \n"
report += f"Best Cross-validation score:  {grid_search.best_score_}\n" 
report += f"Best parameters:  {grid_search.best_params_}\n"

# the performance of the best found parameters on the training and test set
report += f"Training set score with best parameters:  {grid_search.score(X_train, y_train)}\n"
    
print (report)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:680: UserWarning:

The least populated class in y has only 1 members, which is less than n_splits=5.



Gaussian Naive Bayes
Best parameters:  {'var_smoothing': 0.0001}
Training set score with best parameters:  0.20752704230761024



In [ ]:
# possible values for each hyperparameter
param_grid = {"var_smoothing":[1e-09, 1e-08, 1e-07, 1e-06, 1e-05, 1e-04, 1e-03, 0.01, 0.1, 1, 10, 100]}

# apply Grid Search with Gaussian Naive Bayes model and evaluate performance of parameters with 5-fold evaluation
grid_search = GridSearchCV(GaussianNB(), param_grid, cv=5)

#fit the grid search object on the training data training data after feature seletion
grid_search.fit(X=X_train_selected, y=y_train)

report = ''
# result of grid search
report += "Gaussian Naive Bayes cross-validation with selected features\n"
report += f"Best Cross-validation score:  {grid_search.best_score_}\n" 
report += f"Best parameters:  {grid_search.best_params_}\n"

# the performance of the best found parameters on the training and test set
report += f"Training set score with best parameters:  {grid_search.score(X_train_selected, y_train)}\n"
    
print (report)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:680: UserWarning:

The least populated class in y has only 1 members, which is less than n_splits=5.



Gaussian Naive Bayes cross-validation with selected features
Best Cross-validation score:  0.20204375336035843
Best parameters:  {'var_smoothing': 1}
Training set score with best parameters:  0.2008491391265388



Feature selection and hyperparameter tuning enhances our model's performance. 

### 3.5. Model Testing

We now will evaluate our models using the tuned hyperparameters for each model (Decision Tree and Gaussian Naive Bayes). 

In [ ]:
# import GridSearchCV and fit GridSearchCV
from sklearn.model_selection import GridSearchCV

   
# possible values for each hyperparameter
param_grid = {"max_depth":[10, 100, 1000, 10000, 10000],
             "max_leaf_nodes": [10, 100, 1000, 10000, 100000]}

# apply Grid Search with decision tree classifier and evaluate performance of parameters with 5-fold evaluation
grid_search = GridSearchCV(DecisionTreeClassifier(), param_grid, cv=5)

#fit the grid search object on the training data
grid_search.fit(X=X_train, y=y_train)

report = ''
# result of grid search
report += "Decision Tree with tuned parameters\n"

# the performance of the best found parameters on the training and test set
report += f"Training set score with best parameters:  {grid_search.score(X_train, y_train)}\n"
report += f"Testing set score with best parameters:  {grid_search.score(X_test, y_test)}\n"
    
print (report)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:680: UserWarning:

The least populated class in y has only 1 members, which is less than n_splits=5.



Decision Tree with tuned parameters
Training set score with best parameters:  0.2517548164110004
Testing set score with best parameters:  0.2393753200204813



In [ ]:
# possible values for each hyperparameter
param_grid = {"var_smoothing":[1e-09, 1e-08, 1e-07, 1e-06, 1e-05, 1e-04, 1e-03, 0.01, 0.1, 1, 10, 100]}

# apply Grid Search with Gaussian Naive Bayes model and evaluate performance of parameters with 5-fold evaluation
grid_search = GridSearchCV(GaussianNB(), param_grid, cv=5)

#fit the grid search object on the training data 
grid_search.fit(X=X_train, y=y_train)

report = ''
# result of grid search
report += "Gaussian Naive Bayes with tuned parameters \n"

# the performance of the best found parameters on the training and test set
report += f"Training set score with best parameters:  {grid_search.score(X_train, y_train)}\n"
report += f"Testing set score with best parameters:  {grid_search.score(X_test, y_test)}\n"
    
print (report)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:680: UserWarning:

The least populated class in y has only 1 members, which is less than n_splits=5.



Gaussian Naive Bayes with tuned parameters 
Training set score with best parameters:  0.20752704230761024
Testing set score with best parameters:  0.20193292370711727



<a id="4"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

## 4. DISCUSSION



  For our data, we used 3 algorithms in order to compare the results of each one: k Nearest Neighbors, Decision Trees, and Naive Bayes. Before testing classifiers, we first used One Hot Encoding to make our categorical variables easier to manipulate and Standard Scaler to standardize feature variables during data wrangling. On our first attempt, we used the classifiers alongside preliminary training/testing data without hyperparameter tuning. Our results yielded seemingly low accuracy and F1 scores. 

  Following our original attempt, we used recursive feature selection to see if only a subset of feature variables was enough to optimize algorithm performance. Because of this, we could only use non-encoded variables, which ended up being age, episode, rank, and popularity; after feature selection was finished, all features except episode count were chosen as the selected features. We redid the model training and testing and our results yielded similar numbers to our results sans feature selection.

  Out of the 3 algorithms we used, we proceeded to perform hyperparameter tuning on decision trees and Naive Bayes to avoid overfitting our data. We disregarded k Nearest Neighbors as its accuracy score was lower than those of the former two. Between the two, the best algorithm to use would be a decision tree with tuned parameters. These parameters would be max_depth and max_leaf_nodes, both with values of 100. Upon testing our tuned algorithms with the testing set, this confirmed that a Decision Tree would be the best algorithm to use. Our testing set score with the best parameters had a corresponding value of about 0.23. Thus, the decision tree algorithm proved to be the best one for our data. 

  From our findings, however, even with a decision tree algorithm, the scores from grid search on the test data were still very low, with our best score only slightly less than 0.24. Thus, with a decision tree algorithm, even when we factor in only 4 non-encoded variables, the performance would not be very high. So while values for our outcome variables *could* be identified, they would likely not be ideal values and have a very high risk of inaccuracy. However, our model did show improvement after hyperparameter tuning, which leaves room for further improvement.
  
  Ethically, our project's results should not be accepted at face value, simply because there is not much correlation between the features and the target variable. Also, the information from our data source was provided by users who may not have been 100% truthful with their inputs. Therefore, the data may not be as accurate. The introduction of human bias into our data means that there are too many outside factors that have influenced the dataset we used. For future work, it would be interesting to use even more features for our data to continue seeing if there is correlations between these features and our target variable, the overall score. It would also be very interesting if we were to find a certain correlation to actually predict animes that users would want to watch and have it displayed in some sort of user interface such as an app.

<a id="5"></a>
<hr style="height:2px; border:none; color:black; background-color:black;">

### CONTRIBUTIONS


Each member contributed to the report; major decisions were made through group discussions and there was frequent communication throughout the process of the report.

- Caitlin wrote up the data analysis and acquisition processes (2) and summarized our findings (4)
- Rachael did the data wrangling, pre-processing and cleaning (3.1), the model training, optimization and testing (3.3-3.5).
- Michelle wrote up the introduction section (1) and created the visualizations (3.2)
- All group members reviewed and checked each other's work overall.